### PR TITLE
/vg/ Multitool Menu port + More things constructible + AAC

### PR DESCRIPTION
**Pull request by PJB3005**
# Multitool menus port from /vg/
## What is a multitool menu? 
Multitool menus are the menus accessible by multitools on machinery such as Tcomms (saying such as, before this PR it was only Tcomms).
## That sounds sick, what can I do with them?
You can set the following parameters on a variety of machines, currently supported are:
* Vents: Change ID and frequency, which are used in remote control.
* Scrubbers: Change the same things as above.
* Dual-Port air vents: Same as above.
* Injectors: See those injectors in the atmos tanks? Well,you can set the same things as above.
* Gas sensors: Set ID, frequency, and sensor information (Gas composition, pressure, heat)
* Pipe meters: Can set frequency and ID.
* General air control computers (the sensor monitors, GAC is their name in the code): Set frequency and add extra sensors to them.
* Tank control computers: Set frequency, and ID tag of the input and output port of the tank.
* Digital valves: Set ID tag and frequency.

**NOTE:** For some of these, atmospherics/engineering access is required.

# More things constructable.
Due to the fact that these things like ID tags can now be changed, constructability of the following machinery is now possible:
* Air injectors.
* Dual-Port air vents (chances are you never heard of these.)
* Gas sensors.
* GACs (The sensor monitors). Although technically constructable, these would be useless if reconstructed (Board wasn't in the circuit imprinter designs).
* Tank control computers: They now have their own circuit, and as they can now be configured to be made usefull if rebuilt, are constructable.
* Digital valves.

# The AAC: Scripting atmos.

Ever wanted to vent the scrubbers loop *automatically* when the temperature got too high?

Wanted an instant purge on the supply loop when somebody puts plasma in it?

Wait, no?

Well I guess most of you don't play atmos but eh.

Anyways now you can!

# Notes.

**Old PR:** #12 

**Tested and working.**

**Ready for merging.**

### CODE REVIEW

**Comment by Mijhead** [comment]
:+1:

**Comment by TheDZD** [comment]
Just leaving that there until you decide it's ready for merge.

**Comment by tigercat2000** [comment]
:1: :x: :100:

**Comment by PJB3005** [comment]
Welp

I fucked the squash

BRB fixing this

File: code/ATMOSPHERICS/components/unary/vent_scrubber.dm [hunk 1]
```diff
@@ -2,6 +2,8 @@
 	icon = 'icons/atmos/vent_scrubber.dmi'
 	icon_state = "map_scrubber"
 
+	req_one_access_txt = "24;10"
```
Comment (Iamgoofball) [inline code/ATMOSPHERICS/components/unary/vent_scrubber.dm]: why

we removed the access because it was a huge pain in the ass

Comment (PJB3005) [inline code/ATMOSPHERICS/components/unary/vent_scrubber.dm]: I have that there to make multitool menus not work for people without access.

To keep the greytide from fucking with scrubbers and vents in the hallways

Comment (Iamgoofball) [inline code/ATMOSPHERICS/components/unary/vent_scrubber.dm]: uhm

that's completely different to any and all other machines you hack with a multitool

why should vents and scrubbers be special snowflakes about it

**Comment by fj45** [comment]
Thank you so much for porting this once again mate!

Just might want to add a proper changelog.

File: code/ATMOSPHERICS/components/unary/vent_scrubber.dm (hunk 1 above)
Comment (PJB3005) [inline code/ATMOSPHERICS/components/unary/vent_scrubber.dm]: Multitool menus aren't shit like hacking.

They're configuration menus for machinery such as Tcomms

**Comment by Fox-McCloud** [comment]
I'm still not for this, to put it bluntly---I still think it's a lot of needless complexity that only like 2-3 nerds are ever going to use.

That said, I'm not going to utilize my time fighting against it either.

It'll be much less destructive with LINDA, though.

**Comment by Spacemanspark** [comment]
:+1:

File: code/ATMOSPHERICS/components/unary/vent_scrubber.dm (hunk 1 above)
Comment (Iamgoofball) [inline code/ATMOSPHERICS/components/unary/vent_scrubber.dm]: wait, I'm getting this mixed up, im dum, disregard

**Comment by PJB3005** [comment]
Fox, you haven't looked at how much this stuff gets used on /vg/.

Then again you don't have an SM that requires major setup/TeG

**Comment by Spacemanspark** [comment]
I want the SM back.

**Comment by tigercat2000** [comment]
I think more complexity is a good thing. Especially if this includes the buildable airlock stuff, which I believe it does. Even if it doesn't, I'm going to use this for the /vg/ massdriver update.

**Comment by fj45** [comment]
I fully support it.

**Comment by PJB3005** [comment]
Nah, doesn't include buildable cycling airlocks.

**Comment by tigercat2000** [comment]
@PJB3005 It's at least a start, I can probably port that from /vg/ anyways :P

**Comment by PJB3005** [comment]
Yeah though you'll need to come up with something for the docking ports, probably just a readonly ID on the shuttle computer

**Comment by tigercat2000** [comment]
I'm not sure I understand what you mean.

**Comment by Spacemanspark** [comment]
Meaning we could possibly be able to make shuttle docks... and build our own usable shuttles?
Please tell me that's what I'm hearing. That would be amazing.

**Comment by PJB3005** [comment]
building shuttles would be really hard to code & balance.

though, building shuttle docking ports would be possible, but mostly useful for fixing shit after, for example, an explosion

**Comment by tigercat2000** [comment]
aahh I see what you mean, yeah that's probably possible, shuttles already use plaintext variables to link to their docking connectors.

**Comment by PJB3005** [comment]
Yeah, they just use id tags and the radio.

Though even then, using multitool menu built in linking, that wouldn't be hard to get around if it weren't.

**Comment by Regen1** [comment]
#bringbrackSM

:+1:  for this though
Complexity and availability for stuff like this is what makes this game fun, and separates good players from great players

**Comment by tigercat2000** [comment]
@PJB3005 By the way, the atmos automation stuff isn't included in the default compile, you'll need to commit the .dme if you want it activated

**Comment by PJB3005** [comment]
Yeah, I forgot to include that, in the batch of fixes coming up it'll be included

**Comment by PJB3005** [comment]
Almost ready for merging, right now there's an issue where changing ID tag/frequency on a vent/scrubber doesn't remove it from the air alarms, and as such they'll stay in there, though uncontrollable by the air alarms, which might seem laggy or something.

Another issue is that HREF exploits can bypass the access requirements, working on it.

**Comment by ZomgPonies** [comment]
:+1: 
Also I would love to see supermatter make a return, and maybe TeG or RUST or something. Maybe stick on of the new enginers on the engineering outpost or something.

**Comment by tigercat2000** [comment]
@PJB3005 With some of the things on the AAC, you might want to add traceable logs. For example, turning emitters on and off.

**Comment by PJB3005** [comment]
Actually, I think emitters being turned on/off is already being logged by emitters themselves, though I'm not so sure how good that'd function from an admin perspective, I'll see about it

**Comment by tigercat2000** [comment]
@PJB3005 They are logged, but with no key or info attached- just the frequency. But all it would take is deconstructing the console to make it very difficult to track, especially if they were, say, the singulo emitters, with many fingerprints

**Comment by PJB3005** [comment]
Alright, I'll fix that, probably by also sending the prints on the AAC

**Comment by PJB3005** [comment]
that fix doesn't yet work, I'll amend it later

**Comment by SirBloodBuske** [comment]
Yay !

**Comment by PJB3005** [comment]
Alright, Last issue is fixed, ready to merge

**Pull request merged**